ivy.el

```diff
@@ -981,7 +981,19 @@ If the text hasn't changed as a result, forward to `ivy-alt-done'."
                 (if (and ivy--directory
                          (not (eq (ivy-state-history ivy-last)
                                   'grep-files-history)))
-                    (expand-file-name ivy-text ivy--directory)
+                    (if (and (string= ivy-text "")
+                             (eq (ivy-state-collection ivy-last)
+                                 #'read-file-name-internal))
+                        ;; For `read-file-name' compat, unchanged initial input
+                        ;; means that `ivy-read' shall return INITIAL-INPUT.
+                        ;; `read-file-name-default' `string-equal' return value
+                        ;; with provided INITIAL-INPUT to detect that the user
+                        ;; choose the default, `default-filename'.
+                        ;; We must return `ivy--directory' in unchanged form
+                        ;; in cased `ivy--directory' started out as
+                        ;; INITIAL-INPUT in abbreviated form.
+                        ivy--directory  ;unchanged (unexpanded)
+                      (expand-file-name ivy-text ivy--directory))
```

Comment:
Nitpick: I wonder if a `cond` would avoid the need for nested conditionals?
```el
(setf (ivy-state-current ivy-last)
      (cond ((or (not ivy--directory)
                 (eq (ivy-state-history ivy-last) 'grep-files-history))
             ivy-text)
            ((and (string= ivy-text "")
                  (eq (ivy-state-collection ivy-last)
                      #'read-file-name-internal))
             ;; For `read-file-name' compat, unchanged initial input means
             ;; that `ivy-read' shall return INITIAL-INPUT.
             ;; `read-file-name-default' `string-equal' return value with
             ;; provided INITIAL-INPUT to detect that the user choose the
             ;; default, DEFAULT-FILENAME.  We must return `ivy--directory'
             ;; in unchanged form in cased `ivy--directory' started out as
             ;; INITIAL-INPUT in abbreviated form.
             ivy--directory)          ; Unchanged (unexpanded)
            (t
             (expand-file-name ivy-text ivy--directory)))))
(insert (ivy-state-current ivy-last))
```

Reply:
Yup, nice, was more into adding than rewriting at the time.

Reply:
I usually collect minor refactors over time, so I'll just add this to my pile, unless someone beats me to it.